src/eko/output/struct.py

```diff
@@ -0,0 +1,502 @@
+# -*- coding: utf-8 -*-
+import contextlib
+import io
+import logging
+import os
+import pathlib
+import shutil
+import tarfile
+import tempfile
+import time
+import typing
+from dataclasses import dataclass, fields
+from typing import BinaryIO, Dict, Literal, Optional, Tuple
+
+import lz4.frame
+import numpy as np
+import numpy.lib.npyio as npyio
+import yaml
+
+from .. import basis_rotation as br
+from .. import interpolation
+from .. import version as vmod
+
+logger = logging.getLogger(__name__)
+
+PathLike = typing.Union[str, os.PathLike]
+
+
+class DictLike:
+    def __init__(self, **kwargs):
+        pass
+
+    @classmethod
+    def from_dict(cls, dictionary):
+        return cls(**dictionary)
+
+    @property
+    def raw(self):
+        dictionary = {}
+        for field in fields(self):
+            value = getattr(self, field.name)
+
+            # replace numpy arrays with lists
+            if isinstance(value, np.ndarray):
+                value = value.tolist()
+            # replace numpy scalars with python ones
+            elif isinstance(value, float):
+                value = float(value)
+
+            dictionary[field.name] = value
+
+        return dictionary
+
+
+@dataclass
+class Operator(DictLike):
+    alphas: float
+    operator: np.ndarray
+    error: Optional[np.ndarray] = None
+
+    # IO works with streams in memory, in order to avoid intermediate write on
+    # disk (keep read from and write to tar file only)
+
+    def save(self, compress: bool = True) -> Tuple[io.BytesIO, bool]:
```

Comment:
why is stream not coming as argument here? I would expect the method to write to it (think of `<<` in c++)

Reply:
Solved in a8a5818eb007cd1cc3b8815216a7784f001f7e0a